### ISSUE TITLE
support for CTT antenna numbers on SensorStations

### ISSUE THREAD

**Issue by denislepage** [MotusDev/Motus-TO-DO#535]
CTT SensorStations have 2 types of ports: 7 USB ports labeled USB 1 - USB 7, and 5 ports for CTT 433 MHz antennae, labeled L1 - L5.

Our metadata currently doesn't support this, and it is clear that CTT will be unable to reprint their boards anytime soon.

A few options are available, with pros and cons:

1. ask users to manually renumber their ports, and/or use a drop down list that has the same effect.
2. expand the unique requirement so we can have CTT and SG antennae with overlapping port numbers.
3. modify the ant field type so we allow text (e.g. L1). We would keep Lotek/SG USB ports as numeric without prefix, and use L1 to L5 for CTT ports.

In all cases, I also have to apply the same approach to the incoming CTT data coming from SensorStations.

The first option could get very messy. It'll be hard to communicate this to people properly, and if we use a drop-down list to hide this, it will create more problems when they are trying to match the metadata and the data. Once you start changing data, you almost also have to keep track of the changes you make. There is also nothing to say that CTT won't add an 8th USB port later, or add other types of ports, and I am not sure what we would do then.

The second option is also messy. The batches currently have a source field (CTT or SG), and the metadata would need to follow the same approach, so we can merge the data and metadata on the same 2 set of values (ant + source). We'd also have a hard time explaining to users why CTT dongles running on SensorGnome should be sg, not ctt, etc. Why lotek should be sg, etc.

I like the 3rd option by far the most, as it requires almost no tampering with the original data. It'll trigger a cascade of changes in various places, but probably isn't too hard to implement. The biggest constraint is that we'll have to rewrite the large taghits table, but we can implement this in the metadata only to start. It'd be a good time now also to modify the DET table while it's not yet in use, and we can remove a couple superfluous fields at the same time (e.g. nodeNum). Aside from a bit of pain with the implementation, I can't really think of a downside to this approach.

We may need a short term solution until this is ready, and asking people to pick non-overlapping antenna number (e.g. adding 10 to CTT L ports) would make it easy to fix the metadata when we are ready. Or, if we think that we can implement this change simply in beta8 and release within a week or less, we can always tell them to wait a bit.

**Comment by cjardine-bsc** [comment]
I slept on it and couldn't think of any dire consequences to option 3. 

Unless you've thought of anything else I can start implementing the changes in the database.

**Comment by denislepage** [comment]
I think so, but let’s give it a chance for a bit more feedback. We’ll start with sandbox, of course. Make sure you postpone the import queues (CTT-SG and SG. CTT alone is already turned off) before starting.

There will be a few more changes to do in the Java code to switch from a number to a String, I expect.

I’d use varchar(4)? We can use a drop down list to help users, but the values would be :

SensorGnome and Lotek receivers: port number (numeric only)
SensorStation: USB ports (numeric only), CTT ports (L1-L5).

I took a quick look, and there are more tables affected that I realized, but most of it should be straightforward.  I mistakenly thought at first that it was det_taghits that needed to be updated, but of course, it’s det_run (and the backup tables). That’s a much more manageable one to deal with. It will take a while, but it may be possible to rely on SQL to create the alter scripts.

No need to modify the ctt and sg prefixed tables. I look for any tables containing “ant” or “port”. Not sure if I missed anything.

We should check the processes that populate det_taghits_runs and det_taghits_daily and the like to be sure (e.g. temporary tables).

api_import_sensor_antennas
api_import_sensor_gps
bt_api_import_sensor_antennas
bt_api_import_sensor_gps
bt_det_pulse_counts
bt_det_runs
bt_sensor_antennas
#ctt_runs
#ctt_taghits
#ctt_taghits_1
#ctt_taghits_2
det_batches_activity
det_node_data
det_pulse_counts
det_runs
det_taghits_runs
sample_r_batches_activity
sample_r_metadata_for_receivers_antdeps
sample_r_runs_for_tag_project
sensor_antennas
sensor_devices
sensor_gps
sensor_params

Finally, I wonder if the views will need to be recompiled (without any changes). Perhaps.

**Comment by cjardine-bsc** [comment]
Okay, all those changes have been made in Sandbox!

**Comment by denislepage** [comment]
I think we can resume the import queue then. I'll look into it now.

**Comment by pmorrill** [comment]
I am ready to update the interface to accept the varchar ports, but I understood that we were going to add a lk_ table to hold acceptable values (1 - 12, and L1 - L8, for example).

I am going without a dropdown box for the moment, but it would be good to have an acceptable set for verification.

 I don't see that in sandbox - did I miss it?

**Comment by pmorrill** [comment]
I have updated the handler so that it deals with varchar input, and added some simple validation of new port numbers: we only allow strings that cam be interpreted as an Int, or that begin with 'L', followed by an int. If we add a lookup table I will add that extra check.

**Comment by cjardine-bsc** [comment]
You now have a table called lk_ant in sandbox. Enjoy!

**Comment by pmorrill** [comment]
Code to use this lookup is in place; will test in sandbox later this AM, then we can cut a beta9 (I think)

**Comment by Hussell** [comment]
There are a few references to integer port and ant in motus-api ReceiversAPI, SGDataAPI and related procs and views prAPIImportSensorAntennas, prAPIImportSensorGps, vwData_receivers_antennas, vwBase_receivers_antennas, vwR_Metadata_for_receivers_antDeps, vwR_Batches_activity, vwR_Pulse_counts_for_receiver, and probably some I've missed. There are also a bunch in the receiver timeline, which I'm fixing now.

**Comment by pmorrill** [comment]
I will take a look at the api code.

**Comment by cjardine-bsc** [comment]
Thise procs (prAPIImportSensorAntennas etc.) have all be moved to old in sandbox as they are not currently being used for anything. 

The views just need to be dropped and recreated (as Denis mentioned yesterday). I've been working through the DB. I'm doing a compare and update today.

**Comment by Hussell** [comment]
The procs are used in the Java classes mentioned, which are live.

**Comment by pmorrill** [comment]
Part of Motus 2.0 has to be to comb through the old api and remove deprecated entrypoints. The sg server is the only client for that api (ReceiversAPI.java) and it is a headache keeping track of what it uses and what it doesn't. Does it ever register sensors anymore?

In anycase, the 'sensor-deployments list' entry point *does* access the port field from the database, and does use an int to do it. I will do some testing of that and the SG code on Monday.

**Comment by Hussell** [comment]
Yes, sgdata registers every unregistered receiver it encounters. It also fetches a list of all the receiver metadata we have, daily.

**Comment by cjardine-bsc** [comment]
For that, I believe it's using prInsertSensorHashKeys. 

To my knowledge none of the prAPIImportSensors procs are used (at least all the import tables are empty).

We may get to a point where we want to allow sensor deployments to be registered through the API (particularly for CTT) but this, has thus, far not been implemented. I think the procs will need to be re-written if we ever do so. Which is why they have been moved to the old schema and are not being used currently.

**Comment by pmorrill** [comment]
C is right: the sensor registration api call only uses prInsertSensorsHashKeys. I don't believe it ever tries to register ercv deployments, which would involve the port.

The metadata list seems to map as follows:
receiver/list => vwData_receivers
receiver/deployments => vwData_receivers_deployments

But there is another call that needs some work. This one pulls sensors and their deployments using a sql query. I am hoping it is not used now but I may be wrong! Can you can grep through sg code to see if it is used:

receiver/sensordeployments

**Comment by pmorrill** [comment]
In the SG entrypoints (used by R-client), we use 'ant' (which is port id as far as I can see) as an ordering variable. The data records are paginated using this field.... The cases are:

R_Batches_Activity
R_Pulse_counts_for_receiver

Both views are paginated based on 'batchID = ? AND ((ant = ? AND hourbin > ?) OR ant > ?)'

I can easily interpret the incoming ant variable as a String, and feed it into this.... Will there be problems?

**Comment by Hussell** [comment]
receiver/sensordeployments is the access point used daily to refresh the metadata.
receiver/register is the access point used to register receivers.

**Comment by pmorrill** [comment]
ok - thanks. My notes are wrong then. John had sent me a list of access points used and for some reason I have 'receiver/deployments' as the access to get recv metadata. This is helpful.

**Comment by pmorrill** [comment]
so: is the fact that we are going to return a String (Port) in the daily metadata response going to mess up the SG server? I assume he stuffs that data into a table of his own.

**Comment by denislepage** [comment]
We should check, but I can't see a reason why tag finder needs to know about antennae on deployments.

**Comment by denislepage** [comment]
Also, pagination based on the ant field may be ok, as long as the sorting is consistent at both ends (the API and the package). I'm tagging Steffi, so she can check on the R package side. Steffi, we've modified the ant field (same as port in some tables) to support character values instead of only numbers (1, 2, 3, L1, L2, etc). We should change the sqlite data type for the various tables from number to text, and ensure that pagination still works as expected for the couple cases mentioned by Paul.

**Comment by StuMackenzie** [comment]
Setting Frequency on SensorStation Deployments won't stick - eg. set to 166, changed to mulitple, saved, no stick. eg. set to multiple , add new ctt antenna set to 433, save, no changes stuck.

**Comment by pmorrill** [comment]
Stu: a fix is in mast branch for the issue saving frequencies - we'll check it after next build of sandbox, then move to live (beta8) if it holds up. It's a one-liner.

**Comment by steffilazerte** [comment]
So for right now, this breaks the pagination for `activity`. If `ant` is character, pagination fails (infinite loop), if `ant` is numeric pagination proceeds as expected. I can coerce to numeric, but how should I treat the new antennas like "L1"? Should I coerce these to 1? Or leave that to the API?

**Comment by denislepage** [comment]
Antenna 1 and L1 mean different things. A dual-mode receiver will have both, so they need to remain as text throughout. We’re also making changes to the API to support this. I think they are already in place within the sandbox API.

**Comment by steffilazerte** [comment]
Okay, sounds good. @pmorrill Right now pagination with a character/text "ant" value in sandbox doesn't work for the activity API entry ("custom/activity_for_batch"). I assume this is a server side issue, and not one I'll address in R (?)

To test the other api (R_Pulse_counts_for_receiver) I will need someone to give me access to some example receiver data.

**Comment by pmorrill** [comment]
Steffi: I have made an update to the master branch (sandbox) that should fix the pagination issue. Once we next deploy in sandbox, you can revert to using character versions of 'ant' field to test against sandbox, We will put it in beta9 as soon as you give the green-light.

**Comment by Hussell** [comment]
Deployed.

**Comment by steffilazerte** [comment]
Works as expected and I've updated the R package to expect character antenna and port values.

**Comment by steffilazerte** [comment]
I think this is resolved. The `ant` field is indeed a character now. I'd suggest opening a new issue with anything that is still outstanding.
